Domains: Prevent connection of domains that already point to WPCOM during sign-up

leonardost opened the pull request:
## Proposed Changes

- Updates the domain connection process during sign-up, preventing customers from adding a domain connection to a new site if they would need to prove that they own the domain (pcYYhz-kh-p2). A message asking users to purchase a plan to proceed is shown:

<img width="1020" alt="Screenshot 2023-06-06 at 12 47 51" src="https://github.com/Automattic/wp-calypso/assets/5324818/2d5afeff-c69d-4399-810b-ade7867ce4b6">

- The same should happen when trying to connect a domain to am existing free site (i.e. a site without a paid plan). A message asking users to purchase a plan to proceed should be shown:

<img width="1084" alt="Screenshot 2023-06-06 at 12 50 01" src="https://github.com/Automattic/wp-calypso/assets/5324818/19081fd9-8e15-4ea7-a241-7a4405bbc47a">

### Context

Before D112394-code was deployed on 2023-05-31, we weren't checking domain ownership for any domain connections during sign-up, only for existing sites with plans. This means that a bad actor could create a new site with a plan and a domain connection to a domain or subdomain that was previously mapped to us and redirect that domain/subdomain to another, possibly malicious, site.

This PR fixes that problem by requiring users to purchase a plan if the domain connection they want to add requires verification. The connection can be verified and added after the user purchases a plan.

### Screen recordings

- Before:

https://github.com/Automattic/wp-calypso/assets/5324818/a00c9e22-d9f6-4867-bc5e-f4c3a4f09340

- After:

https://github.com/Automattic/wp-calypso/assets/5324818/a9174259-6c74-4ccf-87b9-2b909b53dd11

- After (existing free site without paid plan):

https://github.com/Automattic/wp-calypso/assets/5324818/db010ac7-0861-4094-89c6-c852c63ed8fe

## Testing Instructions

- In your backend, apply these changes 3084d-pb/#diff to disable caching and a11n bypassing
- Open the live Calypso link or build this branch locally
- Navigate to `/start`
- Select the "Use a domain I own" option
- Use a domain that already points to WPCOM (you can use `wpcom3.leotaba.com` if you want)
- Ensure you can't connect that domain during sign-up and a message asking you to buy a plan is shown
- Ensure the rest of sign-up continues as normal, with only a plan being purchased in the end

For an existing free site without a paid plan:
- Ensure you can't connect a domain that requires ownership verification and a message asking you to buy a plan is shown

Also important to test:
- Ensure you can connect a domain that isn't pointing to WPCOM during sign-up
- Try going back and forth along the steps of the sign-up flow and ensure nothing is broken
  - The back button, in particular, should go to different steps if you skipped the domain step or not
  - If you skipped the domain connection step, the plan step should return to the initial domains step
  - If you added a domain connection, the plan step should return to the "transfer or connect" step
- In an existing site, ensure you go through the ownership verification process if you try to connect a domain that points to WPCOM

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

leonardost committed:
Prevent domain ownership verificaiton during sign-up

hambai commented on client/components/domains/use-my-domain/utilities/get-option-info.js:
```diff
@@ -172,6 +172,28 @@ export function getOptionInfo( {
+					sprintf(
+						/* translators: %s - the domain the user wanted to connect */
+						__(
+							"We need to verify you are the owner of <strong>%s</strong> before connecting it, but we're not able to do that during sign-up.<br /><br />Please go back and either choose another domain to connect, or purchase a plan for your site and connect this domain later."
```
What I'm missing here is a CTA link. I htink we can change this sentence to:
> Please complete your plan purchase first in order to connect your domain.
where "complete your plan purchase" is a link/button that skips the domain mapping step and goes to the next step (checkout or whatever it is)

hambai reviewed (changes requested):
As we discussed on our call we need to have a way forward here. I even think we can add a blue button on the right of this seciton that says "Continue" and would cotinue to the next step?

leonardost committed:
Add skip domain selection CTA

leonardost committed:
Fix availability string verification

leonardost committed:
Fix back button after user skips connecting a domain

leonardost committed:
Fix cursor type when hovering link

leonardost committed:
Update CTA button when user cannot conenct a domain in signup

leonardost commented:
Thanks for the review @hambai! I've updated the code with your suggestions, let me know what you think

hambai commented on client/components/domains/use-my-domain/utilities/get-option-info.js:
```diff
@@ -172,6 +173,33 @@ export function getOptionInfo( {
+
+		// We currently aren't handling ownership verification for mapped domains during sign-up
+		// See https://github.com/Automattic/nomado-issues/issues/136 for more context
+		if ( availability.ownership_verification_type !== 'no_verification_required' && isSignupStep ) {
```
I think we should do this not only when isSignupStep is true but also if it's on a site that doesn't have a plan already. Otherwise if you already have the site and go and try to add the mapping it sends you to the auth code verification but it does not check the auth code at all - it just sends you to checkout

leonardost replied:
Good catch! I fixed that now

leonardost committed:
Fix ownership verification on free sites without a plan

If you tried to connect a domain that requried ownership verification
on a free site without a plan, after the ownership verification flow
you'd be taken to the checkout page, regardless if the verification
was successful or not.

hambai reviewed (approved):
Okay, I tested a few diferent cases and it works as expected. Thank's for handling that Leo!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8029633

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.
